docs: Update route conventions (#978)
remix-run/remix: docs/api/conventions.md

- **`app/entry.server.{js,jsx,ts,tsx}`**: This is your entry into the server rendering piece of Remix. This file is required.
- **`app/entry.client.{js,jsx,ts,tsx}`**: This is your entry into the browser rendering/hydration piece of Remix. This file is required.

### Route File Conventions

Setting up routes in Remix is as simple as creating files in your `app` directory. These are the conventions you should know to understand how routing in Remix works.

Please note that you can use either `.jsx` or `.tsx` file extensions depending on whether or not you use TypeScript. We'll stick with `.tsx` in the examples to avoid duplication (and because we ❤️ TypeScript).

#### Root Layout Route

```[3]
app
| routes
| root.tsx
```

> This is your root layout, or "root route" (very sorry for those of you who pronounce those words the same way!). It works just like all other routes:

Reviewer: ```suggestion
The file in `app/root.tsx` is your root layout, or "root route" (very sorry for those of you who pronounce those words the same way!). It works just like all other routes:
```

- You can export a [`loader`](#loader), [`action`](#action), [`meta`](#meta), [`headers`](#headers), or [`links`](#links) function
- You can export an [`ErrorBoundary`](#errorboundary) or [`CatchBoundary`](#catchboundary)
- Your default export is the layout component that renders the rest of your app in an [`<Outlet />`](https://reactrouter.com/docs/en/v6/api#outlet)

#### Basic Routes

```[3-4]
app
| routes
| - about.tsx
| - index.tsx
| root.tsx

> ```

Reviewer: For each of these, could we get another block that shows the URLs this would support? Or maybe even the output of `remix routes`? Could put that in a `<details>` if it's too big. I think that would help a great deal.

Any JavaScript or TypeScript files in the `app/routes/` directory will become routes in your application. The filename maps to the route's URL pathname, except for `index.tsx` which maps to the root pathname.

> The default export in this file is the component that is rendered at that route.

Reviewer: ```suggestion
The default export in this file is the component that is rendered at that route and will render within the `<Outlet />` rendered by the root route.
```

#### Nested Routes

```[3-5]
app
| routes
| - blog
| --- categories.tsx
| --- index.tsx
| - about.tsx
| - index.tsx
| root.tsx
```

Folders inside the `app/routes/` directory will create nested routes and URLs in your app. Files named `index.tsx` will render when the parent layout route's path is matched exactly.

#### Layout Routes

```[3-4]

> app
> | routes
> | - blog
> | --- blog.tsx
> | --- categories.tsx
> | --- index.tsx
> | - about.tsx
> | - index.tsx
> | root.tsx
> ```

Reviewer: I find a really great way to format directory structures is what you get from [`tree`](https://formulae.brew.sh/formula/tree). If you want to just copy/paste some of the characters it uses, check [the jokes tutorial](https://remix.run/docs/en/v1/tutorials/jokes#explore-the-project-structure). For example:

```suggestion
app
├── routes
│   ├── blog
│   │   ├── categories.tsx
│   │   └── index.tsx
│   ├── blog.tsx
│   ├── about.tsx
│   └── index.tsx
└── root.tsx
```

(Note also I think the hierarchy was incorrect because the `blog.tsx` looked like a child of the `blog` directory, so I fixed that as well).

> When a nested route has the same name its directory, it becomes a "layout route" for all of the other child routes inside that directory. Similar to your [root route](#root-layout-route), the layout route should render an `<Outlet />` which is where the child routes will appear. This is how you can create multiple levels of persistent layout nesting associated with URLs.

Reviewer: ```suggestion
In the example above, the `blog.tsx` is a "layout route" for everything within the `blog` directory (`blog/index.tsx` and `blog/categories.tsx`). When a nested route has the same name its directory, it becomes a layout route for all of the other child routes inside that directory. Similar to your [root route](#root-layout-route), the layout route should render an `<Outlet />` which is where the child routes will appear. This is how you can create multiple levels of persistent layout nesting associated with URLs.
```

You can also create layout routes **without adding segments to the URL** by prepending the directory and associated route file with `__`. For example, all of your marketing pages could share a layout rendered in `/app/routes/__marketing.tsx` as the layout, and those routes would go in the `/app/routes/__marketing/` directory. A route `/app/routes/__marketing/product.tsx` would be accessible at the `/product` URL.

> 

Reviewer: ````suggestion

For example:

```
Insert example here
```
````

#### Dot Delimeters

```[8]
app
| routes
| - blog

> | --- blog.tsx

Reviewer: Pretty sure this is supposed to be a sibling to the blog directory

| --- categories.tsx
| --- index.tsx
| - about.tsx
| - blog.authors.tsx
| - index.tsx
| root.tsx
```

By creating a file with `.` characters between segments, you can create a nested URL without nested layouts. For example, a file `/app/routes/blog.authors.tsx` will route to the pathname `/blog/authors`, but it will not share a layout with routes in the `/app/routes/blog/` directory.

#### Dynamic Route Parameters

```[3]
app
| routes
| - blog
| --- $postId.tsx

> | --- blog.tsx

Reviewer: Should be a sibling to the blog directory

| --- categories.tsx
| --- index.tsx
| - about.tsx
| - blog.authors.tsx
| - index.tsx
| root.tsx
```

Routes that begin with a `$` character indicate the name of a dynamic segment of the URL. It will be parsed and passed to your loader and action data as a value on the `param` object.

For example: `/app/routes/blog/$postId.tsx` will match the following URLs:

- `/blog/my-story`
- `/blog/once-upon-a-time`
- `/blog/how-to-ride-a-bike`

> On each of these pages, the dynamic segment of the URL path is the value of the parameter in your route's data.

Reviewer: ````suggestion
On each of these pages, the dynamic segment of the URL path is the value of the parameter. There can be multiple parameters active at any time (as in `/dashboard/:client/invoices/:invoiceId`) and all parameters can be accessed within components via [`useParams`](https://reactrouter.com/docs/en/v6/api#useparams) and within loaders/actions via the argument's [`params`](#loader-params) property:

```tsx
import { useParams } from "remix";
import type { LoaderFunction, ActionFunction } from "remix";

export const loader: LoaderFunction = async ({ params }) => {
  console.log(params.postId);
};

export const action: ActionFunction async ({ params }) => {
  console.log(params.postId);
};

export default function PostRoute() {
  const params = useParams();
  console.log(params.postId);
}
```
````

Also noted: nested routes can contain dynamic segments by using the `$` character in the parent's directory name. For example, `/app/routes/blog/$postId/edit.tsx` might represent the editor view for blog entries.

> See the <Link to="../routing">routing guide</Link> for more information.

Reviewer: Any reason this isn't using a regular markdown link?

```suggestion
See the [routing guide](../routing) for more information.
```

Author: Not sure, that's what it was before but the link was broken anyway 😂 Fixed w/ a regular markdown link.

> <docs-info>Some CLIs require you to escape the \$ when creating files:
> 
> ```bash
> touch routes/\$params.tsx
> ```
> 
> </docs-info>
> 

Reviewer: I don't think enough people will run into this to be worth including in the docs. Not many people use `touch` and those who do will probably know this (or they'll learn pretty quick that it doesn't work and figure it out).

```suggestion
```

#### Splat Routes

```[4,9]
app
| routes
| - blog
| --- $.tsx
| --- $postId.tsx

> | --- blog.tsx

Reviewer: Should be a sibling to the blog directory

| --- categories.tsx
| --- index.tsx
| - $.tsx
| - about.tsx
| - blog.authors.tsx
| - index.tsx
| root.tsx
```

> Files that are named `$.tsx` are called "splat" (or "catch-all") routes. These routes will map to any URL not matched by other route files in the same directory.

Reviewer: Would probably be good to expand this a bit to explain how to access the splat value (I actually don't know how myself 😅).
